etesync-server: add package and configuration for running with uwsgi on nginx [RFC]

peter-stadler opened the pull request:
Maintainer: Peter Stadler <peter.stadler@student.uibk.ac.at>
Compile tested: MIPS 74K, Asus RT-N16, master snapshot
Run tested: MIPS 74K, Asus RT-N16, master snapshot, login for adminstration and connect clients

Description: This includes the main etesync app (django2-etesync-journal) and configuration files to start it with uwsgi on nginx (etesync-uwsgi).

You should be able to login into the admin interface at https://192.168.1.1/etesync/admin (or your router's IP **edited**) using `root` both as username and as password (is there a better way?). For production you should change thereby the password and add other user(s). You can test the android client and the web client https://client.etesync.com/ with any user, also root. In the clients you have to set an arbitrary encryption password and the advanced setting: 
https://192.168.1.1/etesync (or your router's IP **edited**)
 
**NB:** the self signed certificate shows a warning in the browser. The Android client allows to trust the unkown certificate (pin it in the app). <strike>Should we use acme.sh to create the ssl certificate (this needs a real domain)?</strike> (I will not do that)

This is a first proposal and there are some points still missing:

1. <strike>Install etesync to another folder (instead of /etesync-server). What would be the canonical place?</strike> (using `/usr/share/etesync-server`)

2. Maybe add <strike>uci support and</strike> (done) a luci app for etesync, mainly for configuring the server name; there is already a admin interface build in for user management.

3. <strike>Create a self-signed certificate and enable ssl in the etesync server section of the nginx configuration.</strike> (done)

4. <strike>Use a real domain instead of routers IP as servername to make it reachable. I am not sure how to interact with the nginx configuration (right now it is based on my other [PR for nginx](https://github.com/openwrt/packages/pull/11081)). Do you have any suggestions? Would it help to make it available via uci? Or would that be a bad idea or difficult?</strike> (maybe allow later to setup another domain for Nginx, for now make it available only on LAN as default)

5. <strike>Etesync starts a uwsgi instance with the python plugin (this is based on my other [PR for uwsgi](https://github.com/openwrt/packages/pull/9855)); how can I restart it after an update of uwsgi? Would that be a good idea at all?</strike> (using the emperor that is installed by default)

6. <strike>Autostart etesync by: uwsgi --ini /etc/uwsgi/sites/etesync.ini</strike> (done)

7. <strike>The cleanup and preserving files on remove/update is only partial, especially: How do I remove empty folders?</strike> (done)

I appreciate other suggestions for improvement, too. See also the topic at:
https://forum.openwrt.org/t/package-etesync-and-its-dependencies/42269
I hope it is understandable what I try to do.

peter-stadler commented:
This is based on my [other PR for django2 packages](https://github.com/openwrt/packages/pull/9864). Furthermore, it cherry-picks my [PR for nginx: enable /etc/nginx/conf.d directory](https://github.com/openwrt/packages/pull/9859) and my [PR for uwsgi: add package with modules](https://github.com/openwrt/packages/pull/9855).

As @commodo [plans to update his django](https://github.com/openwrt/packages/pull/9864#issuecomment-527070882) package to version 2, I will wait for that and base etesync-server and its other dependencies on his python3-django.

commodo commented:
For reference.
PR is: https://github.com/openwrt/packages/pull/10200

commodo commented on lang/python/sqlparse/Makefile:
```diff
@@ -0,0 +1,34 @@
+PKG_HASH:=7c3dca29c022744e95b547e867cee89f4fce4373f3549ccd8797d8eb52cdb873
+
+PKG_BUILD_DIR:=$(BUILD_DIR)/sqlparse-$(BUILD_VARIANT)-$(PKG_VERSION)
+PKG_UNPACK=$(HOST_TAR) -C $(PKG_BUILD_DIR) --strip-components=1 -xzf $(DL_DIR)/$(PKG_SOURCE)
```
PKG_BUILD_DIR & PKG_UNPACK are useful if you have a `python` variant;
you can remove them if you want

peter-stadler replied:
I removed PKG_UNPACK and edited PKG_BUILD_DIR because without it the build folder would be sqlparse-python3/sqlparse-$(PKG_VERSION) ...

commodo commented on lang/python/sqlparse/Makefile:
```diff
@@ -0,0 +1,34 @@
+  A non-validating SQL parser module. It provides support for parsing, splitting and formatting SQL statements.
+endef
+
+$(eval $(call Py3Package,sqlparse))
```
maybe rename the package to `python3-sqlparse`
sqlparse sounds generic; and there may be other packages that would clash with the name;

commodo replied:
also debian has a `python3-sqlparse` package

peter-stadler replied:
Thank you, I changed it accordingly.

commodo commented on lang/python/sqlparse/Makefile:
```diff
@@ -0,0 +1,34 @@
+include $(TOPDIR)/rules.mk
```
generally speaking, i would split this `sqlparse` package into a separate PR;
it looks pretty ready

peter-stadler replied:
Done [that](https://github.com/openwrt/packages/pull/10375).

commodo commented on lang/python/django2-cors-headers/Makefile:
```diff
@@ -0,0 +1,34 @@
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=django2-cors-headers
```
if you rename `django2` to `django` this looks pretty ready and can be split into another PR

peter-stadler replied:
Thank you, done [that](https://github.com/openwrt/packages/pull/10374).

commodo commented on lang/python/django2-cors-headers/Makefile:
```diff
@@ -0,0 +1,34 @@
+  Django application for handling the server headers required for Cross-Origin Resource Sharing (CORS).
+endef
+
+$(eval $(call Py3Package,django2-cors-headers))
```
i would name this `python3-django-cors-headers`
debian also has this package named the same way

peter-stadler replied:
Thank you again, I changed it accordingly.

peter-stadler commented:
That is perfectly fine. Thank you for the notification. The next week I will have little time, but then I will look into it :-)

peter-stadler commented on lang/python/sqlparse/Makefile:
```diff
@@ -0,0 +1,34 @@
+  CATEGORY:=Languages
+  TITLE:=Non-validating SQL parser module.
+  URL:=https://github.com/andialbrecht/sqlparse
+  DEPENDS:=python3-base 
```
Thank you, changed it to python3-light :-)

peter-stadler commented:
Thanks, I split off the following three python3 packages:

- [sqlparse](https://github.com/openwrt/packages/pull/10375)
- [django-cors-headers](https://github.com/openwrt/packages/pull/10374) including also sqlparse
- [drf-nested-routers](https://github.com/openwrt/packages/pull/10376)

Later I would create another PR for the django-etesync-journal.

peter-stadler commented:
I updated the initial comment (mostly striking the parts done).

peter-stadler commented:
(added `@["ipv6-prefix-assignment"].*["local-address"].address` IPs)

neheb commented on net/etesync-server/files/init.uwsgi:
```diff
@@ -0,0 +1,158 @@
+    option_cb() { return; }
+
+    list_cb() {
+        [ "$1" == "allow_all_ips_of" ] && {
```
== is a bashism. Shellcheck can help here.

peter-stadler replied:
Changed it.

neheb commented on net/etesync-server/Makefile:
```diff
@@ -0,0 +1,143 @@
+#!/bin/sh
+[ -n "$${IPKG_INSTROOT}" ] && exit 0
+rmdir /usr/share/etesync-server/etesync_server
+[ "$${PKG_UPGRADE}" == "1" ] && exit 0
```
== is a bashism.

peter-stadler replied:
changed it :-)

peter-stadler commented:
Thank you, I will look at it in the comming week ...

peter-stadler commented:
I changed it according your suggestions, thank you.

Shellcheck uses `dash` instead of `ash` and complains about string replacement, but that should be [supported](https://www.in-ulm.de/~mascheck/various/ash/#busybox).

When using `#!/bin/sh` it also complains about `local`. Which shebang should I use and should I remove the local keyword?

I removed also the part concerning Nginx (for now at least).

neheb commented:
local keyword can stay. I think @comodo mentioned crashing when it is missing.

peter-stadler commented:
I see, thanks :-)

commodo commented:
> local keyword can stay. I think @commodo mentioned crashing when it is missing.

yep; that was with some older busybox shell; it may still be an issue [i don't know];
if you created too many shell variables, busybox used to run out of some buffer/memory;
but, when using `local`, and group some things with function, env-vars get cleaned up when the function's scope ends;

i don't think it's an issue here; but since `local` is support, i would use it;

> When using `#!/bin/sh` it also complains about `local`. Which shebang should I use and should I remove the local keyword?
maybe use `/bin/sh` as shebang and try to suppress it via shellcheck directives

neheb commented:
@peter-stadler is this ready?

ping @jefferyto

peter-stadler commented:
There is a little problem with the dependency on `python3-django`. On a test install it complains about conflicts with `python3-django1`, although that package is not installed. It goes away, when installing `python3-django` manually beforehand.

I do not see, how I could solve it in this package; I think it's about reverse CONFLICTS (I will write to @jow if it would be good checking for them in `opkg install`).

Beside that I would consider it ready (if using `dash` as shebang is good).

jefferyto commented on net/etesync-server/files/init.uwsgi:
```diff
@@ -0,0 +1,117 @@
+#!/bin/ash /etc/rc.common
```
I think using `/bin/sh` here is fine.

peter-stadler replied:
Perfect, thanks again.

peter-stadler commented on net/etesync-server/files/init.uwsgi:
```diff
@@ -0,0 +1,117 @@
+    [ -e /usr/share/etesync-server/etesync-server.ini ] \
+    || ln -s /var/etc/etesync-server.ini /usr/share/etesync-server/
+
+    cd /etc/uwsgi/vassals/ && ln -s etesync-server.uwsgi etesync-server.ini
```
I think signals should work (also with the emperor) I will look into that :-)

peter-stadler replied:
In the on-demand mode of the emperor stopping is not really working with signals, I am using the ln/touch functionality again, but now on tmpfs; so, the flash will not wear.

tasn replied:
There must be a way to make uwsgi react to signals... Locally I use the `touch etesync-server.ini` method to restart it though. :)

peter-stadler replied:
There is a way you can send SIGINT and co to it (uwsgi --reload and uwsgi --stop are working, too). But, if it is a on-demand vassal of the emperor, it will be reloaded on the next request.
The `touch/ln/rm` mechanism is working also in this setup and it should be no problem anymore as it is now on `/var=tmpfs` (beside it is a bit more complicated in the sense that there is one more symlink).

peter-stadler commented on net/etesync-server/files/init.uwsgi:
```diff
@@ -0,0 +1,117 @@
+    else
+        reset_cb
+
+        echo "$INI" | sed 's/\\n/\n/g' > /var/etc/etesync-server.ini
```
I will do that, thank you.

peter-stadler commented on net/etesync-server/files/init.uwsgi:
```diff
@@ -0,0 +1,117 @@
+
+        if /etc/init.d/nginx running
+        then
+            [ "${NGINX_CHANGED}" -ne 0 ] \
```
Oh, that is a leftover, I will remove it ...

peter-stadler commented on net/etesync-server/files/init.uwsgi:
```diff
@@ -0,0 +1,117 @@
+    config_load etesync_server
+
+    [ -e /usr/share/etesync-server/etesync-server.ini ] \
+    || ln -s /var/etc/etesync-server.ini /usr/share/etesync-server/
```
I see, this would be just to make sure. I will remove it.

peter-stadler commented on net/etesync-server/files/init.uwsgi:
```diff
@@ -0,0 +1,117 @@
+    local name="$2"
+    if [ -n "$type" ] || [ -n "$name" ]
+    then
+        INI="${INI}\n[${name}]\n";
```
Thank you, I will look into it :-)

jefferyto commented:
On the dependency issue:

* This package (`etesync-server`) depends on `python3-django-etesync-journal`, which depends on `python3-django-restframework`, which depends on `django`.
* `django` is not a real package but is provided by `python3-django` and `python3-django1`. (`python3-django` conflicts with `python3-django1` and `python-django1-common`.)
* When considering which `django` package to install for `python3-django-restframework`, opkg is choosing `python3-django1` even though `etesync-server` has a direct dependency on `python3-django`.

I tried moving the `python3-django` dependency ahead of the other Django packages in this package, but opkg still wants to install `python3-django1` (and `python-django1-common` which is a dependency of `python3-django1`).

I'm not really sure what else can be done at this point while keeping the dependencies more or less as they are.

peter-stadler commented on net/etesync-server/Makefile:
```diff
@@ -0,0 +1,137 @@
+		$(1)/usr/share/etesync-server/etesync_server/
+
+	$(INSTALL_DIR) $(1)/www/etesync/static/
+	$(LN) /www/etesync/static $(1)/usr/share/etesync-server/static
```
Thanks, I will use that :-)

peter-stadler commented on net/etesync-server/Makefile:
```diff
@@ -0,0 +1,137 @@
+PKG_RELEASE:=1
+
+PKG_SOURCE:=etesync-server-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://github.com/etesync/server/archive/v$(PKG_VERSION)#
```
Thank you :-)

peter-stadler commented on net/etesync-server/Makefile:
```diff
@@ -0,0 +1,137 @@
+# 	$(call Package/etesync-server)
+# 	TITLE += using gunicorn.
+# 	DEPENDS += +gunicorn3
+# endef
```
In the beginning I was not sure which one would be better, but now I am quite sure that it will be only uwsgi, thank you for pointing me to it. I will merge it :-)

peter-stadler commented on net/etesync-server/Makefile:
```diff
@@ -0,0 +1,137 @@
+
+define Py3Build/Compile
+	$(INSTALL_DIR) $(PKG_INSTALL_DIR)/$(PYTHON3_PKG_DIR)
+endef
```
Thank you :-)

peter-stadler commented on net/etesync-server/Makefile:
```diff
@@ -0,0 +1,137 @@
+
+	$(INSTALL_BIN) $(PKG_BUILD_DIR)/manage.py $(1)/usr/share/etesync-server/
+	$(INSTALL_BIN) $(PKG_BUILD_DIR)/etesync_server/* \
+		$(1)/usr/share/etesync-server/etesync_server/
```
Ah, I see, thank you.

peter-stadler commented on net/etesync-server/Makefile:
```diff
@@ -0,0 +1,137 @@
+rm /usr/share/etesync-server/static
+rm -r /www/etesync/
+exit 0
+endef
```
In /www/etesync should only be static things that will be recreated (right now only the symlink, afterwards the content of /usr/share/etesync-server/static).
The user generated data will be in /usr/share/etesync-server. I am not removing that, especially not the database.
Should I remove only the static folder (and leave the empty folder)?
Or leave also the static files, which are consuming a bit of space?

peter-stadler replied:
It is more than 3MB of static files, so, I will do at least:
```
rm -r /www/etesync/static
rmdir /www/etesync
```
There will be wear-levelling on re-installs (not on upgrades), but I think that is better than consuming the space forever ...

jefferyto replied:
I'm ok to remove it all if is only content that can/will be regenerated.

I see you have added a `static_root` option to the config file... I guess I would prefer if it was either completely managed by the package (directory chosen by the package, created and cleaned up by the package) or if it was completely under the control of the user (location set as a config option, user responsible for managing/deleting it).

peter-stadler replied:
I understand, made it now completely managed by the package.

peter-stadler commented on net/etesync-server/Makefile:
```diff
@@ -0,0 +1,137 @@
+define Package/etesync-server/conffiles
+/etc/config/etesync_server
+/etc/nginx/conf.d/etesync.locations
+/etc/init.d/etesync-server
```
Mhm, thank you :-$

peter-stadler commented on net/etesync-server/files/80_setup-etesync-server:
```diff
@@ -0,0 +1,23 @@
+    python3 manage.pyc migrate --noinput || exit 1
+    python3 manage.pyc shell -c "from django.contrib.auth.models import User; \
+        User.objects.create_superuser('root', '', 'root')"  || exit 1
+    echo "etesync-server: created user 'root' with password 'root'."
```
Yes, this password should be changed (this login is for creating users). Is there a standard way for setup?

jefferyto replied:
I don't think there is a standard way to do setup - something like debconf would be helpful here.

If there is an easy way for users to change this root password later (i.e. through a web interface), I guess it's ok to set a default password during installation (with the understanding that some users may never change the default). Large warnings to tell the user to change the default password would also help :joy:

peter-stadler replied:
I am doing it now with user interaction in this file. Without TTY it is skipped ...

peter-stadler replied:
I think the best would be to force the user to change the password (making it `unusable` at install), I will ask @tasn about it, maybe he knows if this can be done.

tasn replied:
As far as I know, there's no way to force a password change in django without adding extra dependencies and code. Though you probably don't want to go down that path anyway.
What about generating a random password here and printing it for the user to save? Or alternatively as @jefferyto suggested, have them set it up manually?

I can try and think about something tomorrow, but no better ideas come to mind at the moment. What users usually do is just create the superuser manually after installation. Maybe that's another option to consider?

peter-stadler replied:
Thank you very much, I think the best will then be to let the user set it up :-)

tasn replied:
Maybe print out a message about it. All the user needs to do is: `/path/to/manage.py createsuperuser` and they will then be prompted with questions.
Maybe we should add a management command to let users create normal users directly so they don't need to even create a superuser (or use the web admin). Just a thought. What do you think?

peter-stadler replied:
Yes printing the message is good, I will look into that.

I am not sure if I understood you right (management command=manage.py?), I will think about it, but I think this is not really needed. Maybe it would be a nice to have in a larger setup, but on a router it is good if there is a superuser who can create accounts for the family ;-)

tasn replied:
True, you are right. It's a good idea to create a superuser account. Nevermind me. :)

peter-stadler commented on net/etesync-server/files/etesync-server.uwsgi:
```diff
@@ -0,0 +1,29 @@
+; ignore-sigpipe  = true
+log-format=%(method) %(uri) => return %(status) (%(rsize) bytes in %(msecs) ms)
+req-logger = syslog:etesync-uwsgi
+disable-logging = true
```
This logs sensitive data (is an access log), I will look into enabling more error log (without logging user data) ...

peter-stadler replied:
Do you know how to setup Nginx's error log for not logging IPs?

peter-stadler replied:
The logging now is without access_log, I would enable Nginx's error_log if there would not be the client IP.
For uWSGI I enabled more logging: Now it filters only about 20 lines that would be printed on every activation-deactivation-cycle (that will be some times a day when used as on-demand instance).

peter-stadler commented:
I looked into opkg, I think this cannot be handled without changing it, I made a [patch](https://lists.infradead.org/pipermail/openwrt-devel/2020-March/022038.html). As I am not experienced at this, I am glad if you (and/or others) take a look at it ...

peter-stadler commented on net/etesync-server/Makefile:
```diff
@@ -0,0 +1,137 @@
+
+define Package/etesync-uwsgi/conffiles
+$(call Package/etesync-server/conffiles)
+/etc/uwsgi/etesync-server.ini
```
Yes, I meant that, thanks.

neheb commented on net/etesync-server/files/80_setup-etesync-server:
```diff
@@ -0,0 +1,52 @@
+
+cd /usr/share/etesync-server || exit 1
+
+if [ ! -f "db.sqlite3" ]
```
Might be cleaner to do

[ -f "db.sqlite3" ] || { }

peter-stadler replied:
Changed it ...

neheb commented on net/etesync-server/files/uwsgi.init:
```diff
@@ -0,0 +1,135 @@
+        etesync_validate_global etesync_print_global
+        etesync_validate_allowed_hosts etesync_print_allowed_hosts
+        etesync_validate_database etesync_print_database
+    } >"${ETESYNC_INI}"
```
Could move this to the } below and remove the initial { I think.

peter-stadler replied:
Thank you did that.

neheb commented on net/etesync-server/files/uwsgi.init:
```diff
@@ -0,0 +1,135 @@
+
+reload_service() {
+    etesync_init
+    touch /var/etc/etesync-server/uwsgi.ini
```
Guess no SIGHUP support. Unfortunate.

peter-stadler replied:
SIGHUP and/or uwsgi --reload would work, too. But, then we need a pid file for that. And the stop() is not working with SIGINT/uwsgi --stop as on-demand vassal ...

Should I use the signal?

neheb replied:
¯\_(ツ)_/¯

My understanding is faster reload.

peter-stadler replied:
I added it :-)

neheb replied:
Hmmm? Can't procd_set_param pidfile be used?

peter-stadler replied:
The emperor can kill it while idle (and restart it on demand), I don't think we get this into procd easily. Would it be good to have it there?

neheb replied:
Ah never mind then.

peter-stadler replied:
:-)

peter-stadler commented:
For the dependencies it needs to be installed by `opkg install python3-django etesync-server` ...

peter-stadler commented:
(Not logging sigpipe as it leaks the client ip.)

peter-stadler commented:
(Same for writev-errors)

peter-stadler commented:
(minor change adding `uci_` prefix to option `allow_all_ips_of`)

peter-stadler commented:
(rebased)

jefferyto commented on net/etesync-server/Makefile:
```diff
@@ -0,0 +1,106 @@
+define Package/etesync-server/conffiles
+/etc/config/etesync_server
+/etc/nginx/conf.d/etesync.locations
+/usr/share/etesync-server/uwsgi.ini
```
Can this config file be located in `/etc` somewhere? If it a true config file (the user may want to edit and changes be saved across sysupgrades), then the conventional place for it would be in `/etc`. If it isn't meant to be edited by the user, then there is no need to be added here in `conffiles`.

peter-stadler replied:
I will change it, thank you for explaining.

jefferyto commented on net/etesync-server/files/81_setup-etesync-server:
```diff
@@ -0,0 +1,56 @@
+if [ "$(python3 -c "$ETESYNC_HAS_USER_PY3CMD" || exit 1)" = "0" ]
+then
+    echo "===== First, create a superuser of the Webinterface by ====="
+    [ -t 0 ] && python3 manage.pyc createsuperuser ||
```
I'm guessing if the user installs this package in luci, then this will not be called? And the server will still be started by the init script, without a superuser?

peter-stadler replied:
Hmm, I will think again about it, but I did not see a good solution (cf. [above](https://github.com/openwrt/packages/pull/9865#discussion_r388388744)) ...

I could create a superuser with a default password (or a random one) and print  a message about that if the shell is not interactive (and else still call the interactive command). Would that be better?

For now, there is only a message that you have to invoke `python3 manage.pyc createsuperuser`. The server will be started in any case and you can see the login interface at `$LOCALIP/etesync/admin`, but you cannot login without superuser. 

And there are some more steps if you want to reach it from WAN not just locally: you must setup a server in Nginx and add the corresponding domain/ip as allowed_host to the config. There is a [template](https://github.com/etesync/server/blob/master/templates/success.html) available at `$LOCALIP/etesync` linking to the upstream README. I am thinking about:
* using the allowed_host from Nginx configuration where possible
* extending the installed template with some more information
* maybe even creating a LuCI app

But, I think they are better in further PRs …

jefferyto commented on net/etesync-server/files/uci.cfg:
```diff
@@ -0,0 +1,13 @@
+
+config direct 'global'
```
Can `direct` and `indirect` be changed to more meaningful names, or removed entirely? I know they had a function previously but now I don't believe they have a purpose.

peter-stadler replied:
Thank you, I will change it :-)

jefferyto commented on net/etesync-server/files/uwsgi.init:
```diff
@@ -0,0 +1,133 @@
+reload_service() {
+    etesync_init
+    kill -SIGHUP "$(cat "/var/etc/etesync-server/master.pid")" 2>/dev/null ||
+    touch /var/etc/etesync-server/uwsgi.ini
```
```suggestion
    touch -h /var/etc/etesync-server/uwsgi.ini
```

although I'm not sure if modifying the symlink will have the desired effect (of reloading the instance).

peter-stadler replied:
Touching it is a documented method of reloading it. I will test if it works without de-referencing, thanks :-)

peter-stadler replied:
`touch -h` is not available, I will re-create the symlink (that is similar to restarting without etesync_init, but that has no big impact, if it is in on-demand status …)

peter-stadler replied:
I found out, this is not needed at all: If the server is in on-demand mode, it will reload the config anyway (so the SIGHUP for the running server is enough). Sorry for the confusion.

peter-stadler commented:
Done the changes:
* moved `/usr/share/etesync-server/uwsgi.ini` to `/etc/uwsgi/vassals/etesync-server.available`
* removed the line `touch /var/etc/etesync-server/uwsgi.ini` since it is not needed for reloading the server if it is on-demand (else reload it by SIGHUP)
* changed the type in `/etc/config/etesync_server` from direct and indirect to django

and rebased.

jefferyto commented:
Apologies for the delay. Regarding creating the superuser account, I think it would be preferable if the same (or equivalent) thing happens whether the user installs the package from the command line or from luci, but I also think a lot of work has already gone into this PR and it's fine to be merged as it is (any tweaks can be done later).

Thanks for sticking with this @peter-stadler :smile:

peter-stadler commented:
Thank you very much for the excellent guidance. I will think about a good way for LuCI and make another PR then :-)

neheb commented:
I take it this is ready?

peter-stadler commented:
Yes, I think so.
